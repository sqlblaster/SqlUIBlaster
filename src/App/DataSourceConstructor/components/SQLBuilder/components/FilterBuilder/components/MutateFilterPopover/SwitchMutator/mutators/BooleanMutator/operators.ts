enum BooleanOperatorEnum {
  'Is true',
  'Is false',
  'Is empty',
  'Is not empty'
}

export type BooleanOperator = keyof typeof BooleanOperatorEnum;

export const booleanOperators: BooleanOperator[] = [
  'Is true',
  'Is false',
  'Is empty',
  'Is not empty'
];
