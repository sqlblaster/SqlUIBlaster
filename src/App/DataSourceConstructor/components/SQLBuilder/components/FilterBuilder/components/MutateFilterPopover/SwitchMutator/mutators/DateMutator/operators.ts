enum DateOperatorEnum {
  'Previous',
  'Next',
  'Before',
  'After',
  'On',
  'Between',
  'Is empty',
  'Is not empty'
}

export type DateOperator = keyof typeof DateOperatorEnum;

export const dateOperators: DateOperator[] = [
  'Previous',
  'Next',
  'Before',
  'After',
  'On',
  'Between',
  'Is empty',
  'Is not empty'
];
