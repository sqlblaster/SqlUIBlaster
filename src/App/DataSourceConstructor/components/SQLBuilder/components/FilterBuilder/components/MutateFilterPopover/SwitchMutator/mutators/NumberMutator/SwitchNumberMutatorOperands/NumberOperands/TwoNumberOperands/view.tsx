import * as React from 'react';
import { NumberField } from '../NumberField';
import { TwoNumberOperandsViewProps } from './props';

export const TwoNumberOperands: React.FC<TwoNumberOperandsViewProps> = ({
  operands = [0, 0],
  handleOperandChange
}) => (
  // TODO: make proper styling of common styles like padding
  <>
    <div style={{ padding: 10 }}>
      <NumberField
        value={operands[0]}
        onValueChange={handleOperandChange(0)}
      />
      <NumberField
        value={operands[1]}
        onValueChange={handleOperandChange(1)}
      />
    </div>
  </>
);
