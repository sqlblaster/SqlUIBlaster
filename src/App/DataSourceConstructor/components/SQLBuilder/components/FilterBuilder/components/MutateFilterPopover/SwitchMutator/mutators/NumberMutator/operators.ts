enum NumberOperatorEnum {
  'Equal to',
  'Not equal to',
  'Greater than',
  'Greater than or equal to',
  'Less than',
  'Less than or equal to',
  'Between',
  'Is empty',
  'Is not empty'
}

export type NumberOperator = keyof typeof NumberOperatorEnum;

export const numberOperators: NumberOperator[] = [
  'Equal to',
  'Not equal to',
  'Greater than',
  'Greater than or equal to',
  'Less than',
  'Less than or equal to',
  'Between',
  'Is empty',
  'Is not empty'
];
