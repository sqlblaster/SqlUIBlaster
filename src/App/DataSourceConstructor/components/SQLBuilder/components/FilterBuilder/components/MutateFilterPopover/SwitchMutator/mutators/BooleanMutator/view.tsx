import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { CheckedSelect } from '../../CheckedSelect';
import { BooleanOperator, booleanOperators } from './operators';
import { BooleanMutatorViewProps } from './props';
import { booleanMutatorStyles } from './styles';

export const BooleanMutator = withStyles(booleanMutatorStyles)(
  class extends React.Component<BooleanMutatorViewProps> {
    public render() {
      const {
        operator,
        handleOperatorSelect,
        titleComponent,
        classes,
        submitMutationComponent
      } = this.props;

      return (
        <div className={classes.root}>
          <div
            className={ `${classes['title-and-operator']} ${
              classes['common-padding']
            }` }
          >
            <div className={classes.title}>{titleComponent}</div>
            <div className={classes.operator}>
              <CheckedSelect<BooleanOperator>
                selectedItem={operator}
                items={booleanOperators}
                onSelect={handleOperatorSelect}
              />
            </div>
          </div>
          <hr className={classes.hr} />
          <div
            className={ `${classes['mutator-action']} ${
              classes['common-padding']
            }` }
          >
            {submitMutationComponent}
          </div>
        </div>
      );
    }
  }
);
