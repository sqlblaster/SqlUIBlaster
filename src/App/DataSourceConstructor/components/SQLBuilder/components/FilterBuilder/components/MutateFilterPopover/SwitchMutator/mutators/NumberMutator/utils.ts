import { NumberOperands } from './NumberOperation';
import { NumberOperator } from './operators';

export const getDefaultOperands = (
  operator: NumberOperator
): NumberOperands => {
  if (operator === 'Between') {
    return [0, 0];
  } else if (operator === 'Is empty' || operator === 'Is not empty') {
    return undefined;
  } else {
    return 0;
  }
};
