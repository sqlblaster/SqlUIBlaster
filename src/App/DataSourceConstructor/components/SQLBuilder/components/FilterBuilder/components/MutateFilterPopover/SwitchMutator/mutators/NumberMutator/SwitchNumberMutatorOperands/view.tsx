import * as React from 'react';
import { OneNumberOperands } from './NumberOperands/OneNumberOperands';
import { TwoNumberOperands } from './NumberOperands/TwoNumberOperands';
import { SwitchNumberMutatorOperandsViewProps } from './props';

export const SwitchNumberMutatorOperands: React.FC<
  SwitchNumberMutatorOperandsViewProps
> = ({ operator }) => {

  switch (operator) {
    case 'Equal to':
    case 'Not equal to':
    case 'Greater than':
    case 'Greater than or equal to':
    case 'Less than':
    case 'Less than or equal to':
      return <OneNumberOperands />;
    case 'Between':
      return <TwoNumberOperands />;
    case 'Is empty':
    case 'Is not empty':
    default:
      return null;
  }
};
