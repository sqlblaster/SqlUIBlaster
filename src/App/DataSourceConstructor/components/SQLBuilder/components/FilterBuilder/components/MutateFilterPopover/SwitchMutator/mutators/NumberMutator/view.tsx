import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { CheckedSelect } from '../../CheckedSelect';
import { NumberOperator, numberOperators } from './operators';
import { NumberMutatorViewProps } from './props';
import { numberMutatorStyles } from './styles';
import { SwitchNumberMutatorOperands } from './SwitchNumberMutatorOperands';

export const NumberMutator = withStyles(numberMutatorStyles)(
  class extends React.Component<NumberMutatorViewProps> {
    public render() {
      const {
        operator,
        handleSelect,
        titleComponent,
        classes,
        submitMutationComponent
      } = this.props;

      return (
        <div className={classes.root}>
          <div
            className={`${classes['title-and-operator']} ${
              classes['common-padding']
            }`}
          >
            <div className={classes.title}>{titleComponent}</div>
            <div className={classes.operator}>
              <CheckedSelect<NumberOperator>
                selectedItem={operator}
                items={numberOperators}
                onSelect={handleSelect}
              />
            </div>
          </div>
          <hr className={classes.hr} />
          <SwitchNumberMutatorOperands />
          <hr className={classes.hr} />
          <div
            className={`${classes['mutator-action']} ${
              classes['common-padding']
            }`}
          >
            {submitMutationComponent}
          </div>
        </div>
      );
    }
  }
);
