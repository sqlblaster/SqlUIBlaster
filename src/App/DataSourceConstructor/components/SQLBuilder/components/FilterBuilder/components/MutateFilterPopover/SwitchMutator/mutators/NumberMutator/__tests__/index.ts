import {
  fireEvent,
  getAllByDisplayValue as getAllByDisplayValueOn,
  getByDisplayValue as getByDisplayValueOn
} from 'react-testing-library';
import {
  getGeneratedQuery,
  render
} from 'src/App/DataSourceConstructor/components/SQLBuilder/__tests__/render.utils';
import {
  clickAddTheFilter,
  clickUpdateTheFilter
} from 'src/App/DataSourceConstructor/components/SQLBuilder/components/FilterBuilder/__tests__/utils/filter-utils';
import { selectATable } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/TableSelector/__tests__/utils/select-table';
import {
  fireChangeEvent,
  getByText$,
  regex
} from 'src/App/DataSourceConstructor/test-utils/utils';
import { clickAddAFilter } from '../../../../../AddFilter/__tests__/utils/click-add-a-filter';
import { mutatorPanelId } from '../../../../view';
import { NumberOperator, numberOperators } from '../operators';

jest.mock('../SwitchNumberMutatorOperands/NumberOperands/NumberField');

describe(`Number operations`, () => {
  it(`should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  `, async () => {
    const { getByText, getByTestId, debug, componentNode } = render();

    const columns = selectATable();
    const columnName = columns[3].name;

    const clickOnFilter = () => {
      fireEvent.click(getByText(regex(columnName)));
    };

    const checkFilterDescription = (
      operator: NumberOperator,
      additionalDescription: string = ''
    ) => {
      expect(
        getByText$(componentNode, new RegExp(`^${columnName}.+$`, 'i'))
          .textContent
      ).toMatchSnapshot(
        `filter description for '${operator}' number operator${
          additionalDescription ? ` ${additionalDescription}` : ''
        }`
      );
    };

    const checkOperation = (
      nextOperator: NumberOperator,
      prevOperator?: NumberOperator,
      doBeforeSave?: () => void,
      additionalDescription: string = ''
    ) => {
      clickOnFilter();
      prevOperator &&
        fireEvent.click(getByText(new RegExp(`^${prevOperator}.*$`)));
      prevOperator &&
        fireEvent.click(getByText(new RegExp(`^.*${nextOperator}$`)));
      doBeforeSave && doBeforeSave();
      clickUpdateTheFilter();

      checkFilterDescription(nextOperator, additionalDescription);

      expect(getGeneratedQuery()).toMatchSnapshot(
        `'${nextOperator}' number operator${
          additionalDescription ? ` ${additionalDescription}` : ''
        }`
      );
    };

    // default number filter
    clickAddAFilter();
    fireEvent.click(getByText(regex(columnName)));
    clickAddTheFilter();

    checkFilterDescription('Equal to');

    expect(getGeneratedQuery()).toMatchSnapshot(`default number operator`);

    // Set certain operand's value
    clickOnFilter();
    fireChangeEvent(
      getByDisplayValueOn(getByTestId(mutatorPanelId), '0'),
      '123'
    );

    // should not allow non-number operands
    fireChangeEvent(
      getByDisplayValueOn(getByTestId(mutatorPanelId), '123'),
      'FTFR'
    );

    // should allow non-integer operands
    fireChangeEvent(
      getByDisplayValueOn(getByTestId(mutatorPanelId), '123'),
      '123.123'
    );

    clickUpdateTheFilter();

    expect(getGeneratedQuery()).toMatchSnapshot(
      `'Equal to' number operator with certain operand's value`
    );

    // Check all operators
    for (let i = 1; i < numberOperators.length; i++) {
      const nextOperator = numberOperators[i];
      const prevOperator = numberOperators[i - 1];
      checkOperation(nextOperator, prevOperator);
    }

    // Check `Between` operator with certain values
    const setBetweenOperands = (
      operandValue1?: string,
      operandValue2?: string
    ) => {
      const [operand1, operand2] = getAllByDisplayValueOn(
        getByTestId(mutatorPanelId),
        '0'
      );

      operandValue1 && fireChangeEvent(operand1, operandValue1);
      operandValue2 && fireChangeEvent(operand2, operandValue2);
    };

    const lastOperator: NumberOperator =
      numberOperators[numberOperators.length - 1];

    checkOperation(
      'Between',
      lastOperator,
      () => {
        setBetweenOperands(undefined, '5');
      },
      'with second operand set'
    );

    checkOperation(
      'Between',
      undefined,
      () => {
        setBetweenOperands('2', undefined);
      },
      'with both operands set'
    );
  });
});
