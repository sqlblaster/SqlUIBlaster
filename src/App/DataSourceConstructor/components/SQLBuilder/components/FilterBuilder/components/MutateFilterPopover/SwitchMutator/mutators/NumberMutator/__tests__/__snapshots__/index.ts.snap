// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Between' number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" >= 0
  AND \\"orders1\\".\\"Quantity\\" <= 0
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Between' number operator with both operands set 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" >= 2
  AND \\"orders1\\".\\"Quantity\\" <= 5
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Between' number operator with second operand set 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" >= 0
  AND \\"orders1\\".\\"Quantity\\" <= 5
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Equal to' number operator with certain operand's value 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" = 123.123
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Greater than or equal to' number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" >= 0
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Greater than' number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" > 0
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Is empty' number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" IS NULL
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Is not empty' number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" IS NOT NULL
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Less than or equal to' number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" <= 0
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Less than' number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" < 0
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : 'Not equal to' number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" <> 0
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : default number operator 1`] = `
"SELECT
  orders1.*
FROM
  \\"orders\\" \\"orders1\\"
WHERE
  \\"orders1\\".\\"Quantity\\" = 0
LIMIT 2000
"
`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Between' number operator 1`] = `"Quantitybetween0and0"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Between' number operator with both operands set 1`] = `"Quantitybetween2and5"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Between' number operator with second operand set 1`] = `"Quantitybetween0and5"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Equal to' number operator 1`] = `"Quantityis equal to0"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Greater than or equal to' number operator 1`] = `"Quantityis greater than or equal to0"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Greater than' number operator 1`] = `"Quantityis greater than0"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Is empty' number operator 1`] = `"Quantityis empty"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Is not empty' number operator 1`] = `"Quantityis not empty"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Less than or equal to' number operator 1`] = `"Quantityis less than or equal to0"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Less than' number operator 1`] = `"Quantityis less than0"`;

exports[`Number operations should render proper sql query on all operators:
      - with default operand's value
      - with certain operand's value,
      should not allow non-number operands,
      should allow non-integer operands
  : filter description for 'Not equal to' number operator 1`] = `"Quantityis not equal to0"`;
